openedx/core/lib/teams_config.py

```diff
@@ -0,0 +1,412 @@
+        """
+        Get the maximum team size for a cluster, or None if no maximum.
+        """
+        try:
+            cluster = self.clusters[cluster_id]
+        except KeyError:
+            raise ValueError("Cluster '{}' does not exist.".format(cluster_id))
+        if not cluster.management.team_size_limit_enabled:
+            return None
+        if cluster.max_team_size is not None:
+            # Explicitly check against None,
+            # because we do not want to treat Zero as None.
+            return cluster.max_team_size
+        return self.max_team_size
+
+    def to_dict(self):
+        """
+        Return this teams config as JSON-ifyable dictionary.
+        """
+        clusters_key = self.clustering_scheme.value
+        clusters = [cluster.to_dict() for cluster in self.clusters]
+        return {
+            clusters_key: clusters,
+            "max_team_size": self.max_team_size,
+        }
+
+
+class TeamsEnabledWithTopics(TeamsEnabled):
+    """
+    Teams are enabled for a course, and configured to use topics as clusters.
+    """
+    def __init__(self, topics, max_team_size=None, source_data=None):
+        self.topics = topics
+        super(TeamsEnabledWithTopics, self).__init__(
+            max_team_size=max_team_size,
+            source_data=None,
+        )
+
+    @property
+    def clustering_scheme(self):
+        return ClusteringScheme.topics
+
+    @property
+    def clusters(self):
+        return self.topics
+
+
+class TeamsEnabledWithTeamsets(TeamsEnabled):
+    """
+    Teams are enabled for a course, and configured to use teamsets as clusters.
+    """
+    def __init__(self, teamsets, max_team_size=None, source_data=None):
+        self.teamsets = teamsets
+        super(TeamsEnabledWithTeamsets, self).__init__(
+            max_team_size=max_team_size,
+            source_data=None,
+        )
+
+    # Temporarily disable this configuration until it is implemented (MST-9).
+    def is_enabled(self):
+        return False
+
+    @property
+    def clustering_scheme(self):
+        return ClusteringScheme.teamsets
+
+    @property
+    def clusters(self):
+        return self.teamsets
+
+
+class ClusteringScheme(Enum):
+    """
+    The scheme with which the course's teams are divided into clusters.
+    """
+
+    # Under a "topics" scheme, each cluster is a Topic, which generally is a set
+    # teams that are related in subject material.
+    # Students may join ONE TEAM per COURSE.
+    topics = "topics"
+
+    # Under a "teamsets" scheme, each cluster is a Teamset, which generally is a
+    # set of teams formed around an assignment.
+    # Students may join ONE TEAM per TEAMSET.
+    # This scheme allows greater flexibility in that a student may work with
+    # different teams of students for different assignments, all within the
+    # same course.
+    teamsets = "teamsets"
+
+    @classmethod
+    def get_default(cls):
+        return cls.topics
+
+
+class Cluster(object):
```

Comment:
From the Enum description above, it feels like `Clusters` are things that can behave as either a "topic" or a "teamset" - should we have a different class for each, both inheriting from `Cluster`?